packages/datadog-instrumentations/src/aws-sdk.js

```diff
@@ -0,0 +1,103 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+function wrapRequest (send) {
+  return function wrappedRequest (cb) {
+    if (!this.service) return send.apply(this, arguments)
+
+    const serviceIdentifier = this.service.serviceIdentifier
+    const channelSuffix = getChannelSuffix(serviceIdentifier)
+    const startCh = channel(`apm:aws:request:start:${channelSuffix}`)
+    if (!startCh.hasSubscribers) return send.apply(this, arguments)
+    const outerAr = new AsyncResource('apm:aws:request:outer')
+
+    this.on('complete', response => {
+      channel(`apm:aws:request:complete:${channelSuffix}`).publish({ response })
+    })
+
+    return new AsyncResource('apm:aws:request:inner').runInAsyncScope(() => {
+      startCh.publish({
+        serviceIdentifier,
+        operation: this.operation,
+        awsRegion: this.service.config && this.service.config.region,
+        awsService: this.service.api && this.service.api.className,
+        request: this
+      })
+
+      if (typeof cb === 'function') {
+        arguments[0] = wrapCb(cb, channelSuffix, this, outerAr)
+      }
+      return send.apply(this, arguments)
+    })
+  }
+}
+
+function wrapCb (cb, serviceName, request, ar) {
+  return function wrappedCb (err, response) {
+    const obj = { request, response }
+    return ar.runInAsyncScope(() => {
+      channel(`apm:aws:response:start:${serviceName}`).publish(obj)
+      if (!obj.needsFinish) {
```

Comment:
That isn't enough information, because there's nothing to finish if extraction produced nothing. We could just go through the same procedure here regardless, but then we'd need to check for the presence of the maybe-existing span on the subscriber side, and it would be considerably less efficient.